docs/YIP/YIP-0010-state-machine-framework.md

```diff
@@ -0,0 +1,175 @@
+and to make sure that all the edge cases are covered and no security vulnerabilities are introduced.
+
+For this reason, we need to develop simple, but yet extensible framework that allows to create, configure and operate state machines.
+
+## Decision
+
+The state machine framework should stem from the automata theory, given its robustness, proven correctness and wide adoption.
+
+Moreover, given the reactive nature of the trading and settlement protocols, meaning that the workflow starts with one event, which leads to other event generated and so on. The pushdown automaton is the best candidate for the described logic.
+
+### Vocabulary
+
+**Automata theory** - is the study of abstract machines and automata, as well as the computational problems that can be solved using them. It is a theory in theoretical computer science with close connections to mathematical logic.
+
+**Automaton** (comes from the Greek word "αὐτόματος", which means "self-acting, self-willed, self-moving") - is an abstract self-propelled computing device which follows a predetermined sequence of operations automatically.
+
+**State** represents the step of an automaton execution. In the context of this YIP we will only consider automata with finite number of states, which are called finite automaton (FA) or finite state machines (FSM).
+
+**Symbol** - is an input that triggers the transition from one state to another.
+
+**Alphabet** - is a finite set of symbols.
+
+**Transition** - is a function that maps a state and an input symbol to the next state.
+
+**Initiat states** - is a set of states that the automaton can be in before any input.
+
+**Final states** - is a set of states that the automaton can be in after the input is processed.
+
+**Run** - is a sequence of states and input symbols that the automaton goes through during the processing of a string (a set of symbols).
+
+The automaton **halts** when there are no more symbols in the input string.
+
+**Complete automaton** - is an automaton that has a transition for every state and every symbol in the alphabet.
+
+**Deterministic automaton** - is an automaton that has at most one transition for every state and every symbol in the alphabet, and has no epsilon transitions (transitions that do not consume any input).
+
+### Pushdown automaton
+
+**Pushdown automaton** (PDA) - is a finite automaton that employs a stack as a memory to store an arbitrary number of symbols from the alphabet of the PDA, but allows reading only the topmost symbol.
+
+The term "pushdown" refers to the fact that the stack can be regarded as being "pushed down" like a tray dispenser at a cafeteria, since the operations never work on elements other than the top element. A stack automaton, by contrast, does allow access to and operations on deeper elements.
+
+### State machine framework
+
+The state machine framework shall be implemented as a library that can be used by the trading and settlement protocols.
```

Comment:
clearing and settlement

Reply:
16adb4568b819f92525fd671c4abfea360311f49